### PR TITLE
Crank-Nicolson Leapfrog (CNLF)

### PR DESCRIPTION
**Pull request by sipah00**

### CODE REVIEW

**Comment by sipah00** [comment]
Getting `convergence test` fail when `f1` is not zero, also `sol` is far away from actual sol. It only works when `f1` is zero.

**Comment by sipah00** [comment]
@ChrisRackauckas could you look once, where am I doing mistake.

**Comment by ChrisRackauckas** [comment]
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/386/files#diff-1cc7ad80e3fb1d3afc1680b3a07c2ca3R1679

that part is incorrect. You need to skip over `k2` for one round. So you need two caches in order to have it wait one iteration before being reused.

**Comment by sipah00** [comment]
why? we need value of `f1` at `t-dt`.

**Comment by ChrisRackauckas** [comment]
Oh I see what you're doing. Clever 😄 . Yeah, I'm not sure of the issue. But I do see something else. Slack chat.

**Comment by sipah00** [comment]
@ChrisRackauckas In `CNLF2`, we can not use leapfrog method for first step because we need value at the previous point

**Comment by ChrisRackauckas** [comment]
Yeah, so do implicit Euler for the implicit part in the first step. Then, when the first step is at its end, update the historical value to be `f1(uprev,p,t)`, so then it's `t-dt` in the next step.

**Comment by sipah00** [comment]
and for explicit part, should I use `explicit euler` ?

**Comment by ChrisRackauckas** [comment]
Yes, for the explicit part do explicit Euler. Details for a simplified implementation of it:

make the `t-dt` term 0
and double the `t+dt` term
that makes the `g` part equivalent to implicit Euler and the implicit code is all the same in this step except for the `gamma` coefficient.

Then the explicit term will need a bit more handling to switch from explicit Euler to the Leapfrog, but this will just require a quick loop.

**Comment by sipah00** [comment]
cnlf1
`blue` -> `Tsit5`
`Orange` -> `ABCN2`
`green` -> `CNLF2`
this is what I was talking about, `sol` is far away from actual sol (green (CNLF2)).
```julia
f1 = (u,p,t) -> 2u
f2 = (u,p,t) -> 2u
prob = SplitODEProblem(f1,f2,1.0,(0.0,1.0))
sol = solve(prob,CNLF2(),dt=1//256)
```

**Comment by ChrisRackauckas** [comment]
The tests do the two parts separate. Which part is converging incorrectly? The explicit or implicit part?

**Comment by sipah00** [comment]
implicit

**Comment by sipah00** [comment]
okay

**Comment by sipah00** [comment]
no improvement

File: src/perform_step/adams_bashforth_moulton_perform_step.jl [hunk 1]
```diff
@@ -1608,32 +1607,86 @@ function perform_step!(integrator, cache::ABCN2Cache, repeat_step=false)
...
+  if cnt == 1
+    tmp = uprev + dt * (integrator.fsalfirst - du₁)
+  else
+    tmp = uprev2 + 2 * dt * (integrator.fsalfirst - du₁)
   end
-  integrator.f(k,u,p,t+dt)
+  # Implicit part
+  # precalculations
+  if cnt == 1
+    γ = 1
```
Comment (ChrisRackauckas) [inline src/perform_step/adams_bashforth_moulton_perform_step.jl]: this isn't type-stable. Make it `1//1`. There's probably a similar change that should be done in other places.

**Comment by ChrisRackauckas** [comment]
Then I would start printing out values. You know that the `k1` and `k2` are derivative calculations, so for the linear ODE they should be `alpha * u`. Check to make sure they all line up with the correct values. Check if `W = I - gamma*dt*J` is correct.

**Comment by sipah00** [comment]
yeah, I will start checking every value with theoretical for a test example.

**Comment by sipah00** [comment]
```julia
f1 = (u,p,t) -> 2u
f2 = (u,p,t) -> 2u
prob = SplitODEProblem(f1,f2,1.0,(0.0,1.0))
sol = solve(prob,CNLF2(),dt=0.2)
```
I solved this problem using pen and paper, and compared with values given by code, and both are same.

**Comment by sipah00** [comment]
@YingboMa this method is not performing good, see above plot

**Comment by ChrisRackauckas** [comment]
Can you show the plot with f1=0, and the plot with f2=0?

**Comment by sipah00** [comment]
slack chat

**Comment by sipah00** [comment]
ohh.. I was just going to push mutable cache method.

**Comment by sipah00** [comment]
will open new PR.

**Comment by ChrisRackauckas** [comment]
oh shit, I missed that that one was missing. Just do it in a new PR

**Pull request merged**